pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,259 @@
+	stopCh          chan struct{}
+	indexer         cache.Indexer
+	lister          cache.GenericLister
+	gvkTrans        *gvk.GVKTranslator
+}
+
+func (c *Controller) enqueueResource(gvr schema.GroupVersionResource, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited(gvr.String() + "::" + key)
+}
+
+func (c *Controller) enqueueNamespace(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited("namespace::" + key)
+}
+
+func (c *Controller) enqueueCluster(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited("cluster::" + key)
+}
+
+func (c *Controller) Start(numThreads int) {
+	defer c.queue.ShutDown()
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(c.startWorker, time.Second, c.stopCh)
+	}
+	klog.Infof("Starting workers")
+	<-c.stopCh
+	klog.Infof("Stopping workers")
+}
+
+func (c *Controller) startWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	var err error
+	if strings.HasPrefix(key, "namespace::") {
+		err = c.processNamespace(strings.TrimPrefix(key, "namespace::"))
+	} else if strings.HasPrefix(key, "cluster::") {
+		err = c.processCluster(strings.TrimPrefix(key, "cluster::"))
+	} else {
+		err = c.process(key)
+	}
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	// Re-enqueue up to 5 times.
+	num := c.queue.NumRequeues(key)
+	if num < 5 {
+		klog.Errorf("Error reconciling key %q, retrying... (#%d): %v", key, num, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	// Give up and report error elsewhere.
+	c.queue.Forget(key)
+	runtime.HandleError(err)
+	klog.Infof("Dropping key %q after failed retries: %v", key, err)
+}
+
+// key is gvr::namespace/name
+func (c *Controller) process(key string) error {
+	parts := strings.SplitN(key, "::", 2)
+	if len(parts) != 2 {
+		return fmt.Errorf("invalid key: %s", key)
+	}
+	gvrstr := parts[0]
+	gvr, _ := schema.ParseResourceArg(gvrstr)
+	if gvr == nil {
+		return fmt.Errorf("Error parsing GVR %q", gvrstr)
+	}
+	key = parts[1]
+
+	obj, exists, err := c.indexer.GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		klog.Infof("Object with key %q was deleted", key)
+		return nil
+	}
+	current := obj.(*unstructured.Unstructured)
+	previous := current.DeepCopy()
+
+	ctx := context.TODO()
```

Comment:
Don't we have one of these we can pass in to the constructor?

Reply:
Done.